### PR TITLE
ARROW-3384: [Gandiva][C++] Sync remaining commits from Gandiva.

### PR DESCRIPTION
**Pull request by praveenbingo**
I've synced the rest of the commits from Gandiva.
Fixed all of the compilation and style warnings.

### CODE REVIEW

File: cpp/src/gandiva/llvm_generator_test.cc [hunk 1]
```diff
@@ -179,7 +179,7 @@ TEST_F(TestLLVMGenerator, TestNullInternal) {
       reinterpret_cast<uint8_t*>(&local_bitmap),
   };
 
-  eval_func(addrs, local_bitmap_addrs, num_records);
+  eval_func(addrs, local_bitmap_addrs, 0, num_records);
```
Comment (wesm) [inline cpp/src/gandiva/llvm_generator_test.cc]: What are the 0's here?

File: java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java [hunk 2]
```diff
@@ -752,6 +871,67 @@ public void testTimeNull() throws GandivaException, Exception {
     eval.close();
   }
 
+  @Test
+  public void testGDV117() throws GandivaException, Exception {    /*
```
Comment (wesm) [inline java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java]: Uses a private JIRA id

**Review by wesm — COMMENTED** [review]
Some minor comments.

**Comment by wesm** [comment]
To other maintainers -- leave this to me to merge since I want to tweak the commit messages a bit (add spaces after "[Gandiva]" for readability, edit the message on the patch with no "[Gandiva]")

File: cpp/src/gandiva/compiled_expr.h [hunk 3]
```diff
@@ -23,7 +23,8 @@
 
 namespace gandiva {
 
-using EvalFunc = int (*)(uint8_t** buffers, uint8_t** local_bitmaps, int record_count);
+using EvalFunc = int (*)(uint8_t** buffers, uint8_t** local_bitmaps,
+                         int64_t execution_ctx_ptr, int record_count);
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/compiled_expr.h]: fixed..

File: cpp/src/gandiva/date_utils.cc [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+namespace gandiva {
+
+#ifdef GDV_HELPERS
+namespace helpers {
+#endif
+
+std::vector<std::string> DateUtils::GetMatches(std::string pattern, bool exactMatch) {
+  // we are case insensitive
+  std::transform(pattern.begin(), pattern.end(), pattern.begin(), ::tolower);
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/date_utils.cc]: yes the pattern would be ascii only..

File: cpp/src/gandiva/expr_decomposer_test.cc [hunk 5]
```diff
@@ -43,6 +43,9 @@ TEST_F(TestExprDecomposer, TestStackSimple) {
   // else _
   IfNode node_a(nullptr, nullptr, nullptr, int32());
 
+  decomposer.PushConditionEntry(node_a);
+  decomposer.PopConditionEntry(node_a);
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/expr_decomposer_test.cc]: looks like it it to simulate/test what happens when the visit method is called directly..

File: cpp/src/gandiva/filter_cache_key.h [hunk 6]
```diff
@@ -56,18 +58,44 @@ class FilterCacheKey {
...
 
   bool operator!=(const FilterCacheKey& other) const { return !(*this == other); }
 
   SchemaPtr schema() const { return schema_; }
 
+  std::string ToString() const {
+    std::stringstream ss;
+    ss << "Schema [";
+
+    // remove newlines from schema
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/filter_cache_key.h]: sure done

File: cpp/src/gandiva/llvm_generator.cc [hunk 7]
```diff
@@ -181,8 +186,8 @@ llvm::Value* LLVMGenerator::GetLocalBitMapReference(llvm::Value* arg_bitmaps, in
 // IR Code
 // --------
 //
-// define i32 @expr_0(i64* %args, i64* %local_bitmaps, i32 %nrecords) {
-// entry:
+// define i32 @expr_0(i64* %args, i64* %local_bitmaps, i64 %execution_context_ptr, , i32
+// %nrecords) { entry:
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/llvm_generator.cc]: sure, changed that to a int64

File: cpp/src/gandiva/like_holder.h [hunk 8]
```diff
@@ -49,6 +52,9 @@ class LikeHolder : public FunctionHolder {
 
   std::string pattern_;  // posix pattern string, to help debugging
   RE2 regex_;            // compiled regex for the pattern
+
+  static RE2 starts_with_regex_;  // pre-compiled pattern for matching starts_with
+  static RE2 ends_with_regex_;    // pre-compiled pattern for matching ends_with
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/like_holder.h]: yes [https://github.com/google/re2/blob/master/re2/re2.h#L256](url)

File: cpp/src/gandiva/tests/projector_test.cc [hunk 9]
```diff
@@ -522,4 +668,88 @@ TEST_F(TestProjector, TestZeroCopyNegative) {
...
+
+  // output fields
+  auto field_div = field("divide", int32());
+
+  // Build expression
+  auto div_expr = TreeExprBuilder::MakeExpression("divide", {field0, field1}, field_div);
+
+  std::shared_ptr<Projector> projector;
+  Status status = Projector::Make(schema, {div_expr}, &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/tests/projector_test.cc]: agreed, will fix as part of https://issues.apache.org/jira/browse/ARROW-3358

File: cpp/src/gandiva/llvm_generator_test.cc (hunk 1 above)
Comment (praveenbingo) [reply cpp/src/gandiva/llvm_generator_test.cc]: added a comment.

File: java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java (hunk 2 above)
Comment (praveenbingo) [reply java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java]: fixed.

File: cpp/src/gandiva/tests/if_expr_test.cc [hunk 10]
```diff
@@ -256,6 +256,70 @@ TEST_F(TestIfExpr, TestNestedInIf) {
...
+
+  // Create a row-batch with some sample data
+  int num_records = 6;
+  auto array_a =
+      MakeArrowArrayInt32({21, 15, 5, 22, 15, 5}, {true, true, true, true, true, true});
+  auto array_b = MakeArrowArrayInt32({20, 18, 19, 20, 18, 19},
+                                     {true, true, true, false, false, false});
+  // expected output
+  auto exp =
+      MakeArrowArrayInt32({1, 2, 2, 3, 3, 3}, {true, true, true, true, true, true});
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/tests/if_expr_test.cc]: agreed,we are already doing a story to make it easier to generate an expression from a string, we can do something similar for data..

File: cpp/src/gandiva/projector_cache_key.h [hunk 11]
```diff
@@ -30,16 +32,18 @@ class ProjectorCacheKey {
...
     static const int kSeedValue = 4;
     size_t result = kSeedValue;
     for (auto& expr : expression_vector) {
       std::string expr_as_string = expr->ToString();
       expressions_as_strings_.push_back(expr_as_string);
       boost::hash_combine(result, expr_as_string);
+      UpdateUniqifier(expr_as_string);
     }
     boost::hash_combine(result, configuration);
     boost::hash_combine(result, schema_->ToString());
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/projector_cache_key.h]: as of now no, same expression with different names will be a new compilation.

opening a issue to investigate more.

File: cpp/src/gandiva/date_utils.cc [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+  return matches;
+}
+
+std::vector<std::string> DateUtils::GetPotentialMatches(std::string pattern) {
+  return GetMatches(pattern, false);
+}
+
+std::vector<std::string> DateUtils::GetExactMatches(std::string pattern) {
+  return GetMatches(pattern, true);
+}
```
[the remark this answers is not in the document]
Comment (praveenbingo) [reply cpp/src/gandiva/date_utils.cc]: fixed.

File: cpp/src/gandiva/filter_cache_key.h [hunk 13]
```diff
@@ -56,18 +58,41 @@ class FilterCacheKey {
...
 
   bool operator!=(const FilterCacheKey& other) const { return !(*this == other); }
 
   SchemaPtr schema() const { return schema_; }
 
+  std::string ToString() const {
+    std::stringstream ss;
+    // indent, window, indent_size, null_rep and skip new lines.
+    arrow::PrettyPrintOptions options{0, 10, 2, "null", true};
+    PrettyPrint(*schema_.get(), options, &ss);
```
Comment (wesm) [inline cpp/src/gandiva/filter_cache_key.h]: Surprised this does not cause a compiler warning

File: cpp/src/gandiva/projector_cache_key.h [hunk 14]
```diff
@@ -58,18 +62,54 @@ class ProjectorCacheKey {
...
 
   bool operator!=(const ProjectorCacheKey& other) const { return !(*this == other); }
 
   SchemaPtr schema() const { return schema_; }
 
+  std::string ToString() const {
+    std::stringstream ss;
+    // indent, window, indent_size, null_rep and skip new lines.
+    arrow::PrettyPrintOptions options{0, 10, 2, "null", true};
+    PrettyPrint(*schema_.get(), options, &ss);
```
Comment (wesm) [inline cpp/src/gandiva/projector_cache_key.h]: Same re: compiler warning

**Review by wesm — APPROVED** [review]
+1. I'll merge after the build runs

**Comment by wesm** [comment]
Just normalized the commit messages and added spaces for readability

**Comment by wesm** [comment]
+1

**Comment by wesm** [comment]
Merged. Thank you!

**Pull request closed without merging**